pkg/controllers/oauthendpoints/oauth_endpoints_controller.go

```diff
@@ -0,0 +1,273 @@
+	kubeInformersForTargetNS informers.SharedInformerFactory,
+	recorder events.Recorder,
+) factory.Controller {
+	endpointsListFunc := func() ([]string, error) {
+		return listOAuthServices(kubeInformersForTargetNS.Core().V1().Services().Lister(), recorder)
+	}
+
+	getTLSConfigFunc := func() (*tls.Config, error) {
+		return getOAuthEndpointTLSConfig(kubeInformersForTargetNS.Core().V1().ConfigMaps().Lister(), recorder)
+	}
+
+	return endpointaccessible.NewEndpointAccessibleController(
+		"OAuthServerService",
+		operatorClient,
+		endpointsListFunc, getTLSConfigFunc,
+		[]factory.Informer{
+			kubeInformersForTargetNS.Core().V1().ConfigMaps().Informer(),
+			kubeInformersForTargetNS.Core().V1().Services().Informer(),
+		},
+		recorder)
+}
+
+// NewOAuthServiceEndpointsCheckController returns a controller that checks the health of authentication service
+// endpoints.
+func NewOAuthServiceEndpointsCheckController(
+	operatorClient v1helpers.OperatorClient,
+	kubeInformersForTargetNS informers.SharedInformerFactory,
+	recorder events.Recorder,
+) factory.Controller {
+	endpointsListFn := func() ([]string, error) {
+		return listOAuthServiceEndpoints(kubeInformersForTargetNS.Core().V1().Endpoints().Lister(), recorder)
+	}
+
+	getTLSConfigFunc := func() (*tls.Config, error) {
+		return getOAuthEndpointTLSConfig(kubeInformersForTargetNS.Core().V1().ConfigMaps().Lister(), recorder)
+	}
+
+	return endpointaccessible.NewEndpointAccessibleController(
+		"OAuthServerServiceEndpoints",
+		operatorClient,
+		endpointsListFn, getTLSConfigFunc,
+		[]factory.Informer{
+			kubeInformersForTargetNS.Core().V1().Endpoints().Informer(),
+			kubeInformersForTargetNS.Core().V1().ConfigMaps().Informer(),
+		},
+		recorder)
+}
+
+func listOAuthServiceEndpoints(endpointsLister corev1listers.EndpointsLister, recorder events.Recorder) ([]string, error) {
+	var results []string
+	endpoints, err := endpointsLister.Endpoints("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		recorder.Warningf("OAuthServiceEndpointsCheck", "failed to get oauth service endpoints: %v", err)
+		return results, nil
+	}
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			for _, port := range subset.Ports {
+				results = append(results, net.JoinHostPort(address.IP, strconv.Itoa(int(port.Port))))
+			}
+		}
+	}
+	if len(results) == 0 {
+		return nil, fmt.Errorf("oauth service endpoints are not ready")
+	}
+	return toHealthzURL(results), nil
+}
+
+func listOAuthServices(serviceLister corev1listers.ServiceLister, recorder events.Recorder) ([]string, error) {
+	var results []string
+	service, err := serviceLister.Services("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		recorder.Warningf("OAuthServiceCheck", "failed to get oauth service: %v", err)
+		return nil, err
+	}
+	for _, port := range service.Spec.Ports {
+		results = append(results, net.JoinHostPort(service.Spec.ClusterIP, strconv.Itoa(int(port.Port))))
```

Comment:
should we take only ports with `https` name?

Reply:
I'd be inclusive